core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/interceptor/RequestCompressionInterceptor.java

```diff
@@ -85,8 +86,19 @@ public Optional<RequestBody> modifyHttpContent(Context.ModifyHttpRequest context
         Compressor compressor = resolveCompressionType(executionAttributes);
         RequestBody requestBody = context.requestBody().get();
 
-        // TODO - Sync streaming compression implementation
-        throw new UnsupportedOperationException();
+        if (isTransferEncodingChunked(context)) {
+            InputStream compressedStream = compressor.compress(requestBody.contentStreamProvider().newStream());
+            try {
+                byte[] compressedBytes = IoUtils.toByteArray(compressedStream);
+                return Optional.of(RequestBody.fromBytes(compressedBytes));
+            } catch (IOException e) {
+                throw SdkClientException.create(e.getMessage(), e);
+            }
+        }
+
+        CompressionContentStreamProvider streamProvider =
+            new CompressionContentStreamProvider(requestBody.contentStreamProvider(), compressor);
+        return Optional.of(RequestBody.fromContentProvider(streamProvider, requestBody.contentType()));
```

Comment:
It looks like this enables chunked encoding? Are we missing the `Transfer-Encoding: chunked` header? Are we making sure that there is no `Content-Length` header?